akka-stream/src/main/scala/akka/stream/impl/Timers.scala

```diff
@@ -138,13 +138,21 @@ object Timers {
 
   }
 
-  final class IdleTimeoutBidi[I, O](val timeout: FiniteDuration) extends GraphStage[BidiShape[I, I, O, O]] {
+  /**
+   * @param timeoutToTermination allows customising the behaviour once a timeout triggers.
+   *                             Returning None causes completion, and Some(exception) causes failure of the stage with that exception.
+   */
+  final class IdleTimeoutBidi[I, O](val timeout: FiniteDuration, timeoutToTermination: () ⇒ Option[Throwable]) extends GraphStage[BidiShape[I, I, O, O]] {
```

Comment:
> I don't think so - we have a clear difference in "throwing is logged, erroring is not logged automatically". Changing this in recover would be inconsistent.

I don't really get what do you say here. The timer stages only log failures themselves if there is a throw, not a failStage(). If that is the case, then it is a bug and should be changed to failStage. Using recover down the line allows to change the error into a normal completion. But again, I don't see why this is needed. As far as I see (given that the timer stage itself does not throw, hence don't log) the only error logging is at the TCP sink side, which is only on DEBUG level which is the correct thing to do.